### DISCUSSION TITLE
How to retrieve some pod ip ?

### DISCUSSION THREAD

**Discussion by EricBoix** [argoproj-labs/hera#580] [Q&A] [answered]
Consider the following workflow

```python
# Hera version 4.4.2
from hera import (Task, Workflow)

with Workflow("retrieve-deamon-pod-ip-", generate_name=True) as w:

  daemon_t = Task(
        "startpostgresdb",
        daemon=True,
        image="postgres:15.2",
  )
  consumer_t = Task(
        "isdbready",
        image="postgres:15.2",
        command=["/bin/bash", "-c"],
        # Following line will trigger an "Internal Server Error" on submission
        # with error message
        #          templates.isdbready: failed to resolve {{tasks.startpostgresdb.ip}}
        args=["exec pg_isready -U user -h " + daemon_t.ip + " -p 5432"],
  )
  daemon_t >> consumer_t 

w.create()
```

will trigger on submission the error `templates.isdbready: failed to resolve {{tasks.startpostgresdb.ip}}`.
What am I doing wrong ?

Note: when using AW the syntax `{{steps.startpostgresdb.ip}}` (that is `steps...` as opposed to `tasks...`) was the one required for the workflow to be valid.

**Comment by flaviuvadan** [comment, marked as answer]
Thanks for the question @EricBoix! Any chance you looked at the [daemon example](https://github.com/argoproj-labs/hera/blob/v4/examples/daemon.py#L38)? It has a sample usage of the task IP exposed via an env variable!

**Comment by EricBoix** [comment]
Thanks @flaviuvadan. Yes, I did encounter the [deamon example](https://github.com/argoproj-labs/hera/blob/v4/examples/daemon.py#L38) that I could run smoothly while using the `<some-task-name>.ip` syntax.

Indeed for the example I took in this discussion, flowing the `<some-task-name>.ip` through the environment (the way [deamon example](https://github.com/argoproj-labs/hera/blob/v4/examples/daemon.py#L38) does it) would work. But this is only because the [pg_isready SQL related command](https://www.postgresql.org/docs/current/app-pg-isready.html) accepts the [environment variables supported by libpq](https://www.postgresql.org/docs/current/libpq-envars.html) (which includes `PGHOST`). 

But some of my use cases (arising from yaml AW e.g. [this workflow](https://github.com/VCityTeam/ExpeData-Workflows_testing/blob/master/ArgoWorkflows/Workflows_In_Yaml/CityGMLto3DTiles_Example/full-workflow.yml#L168)) are a bit different. They do require that`<some-task-name>.ip` be given/transmitted as `inputs` or `arguments` as opposed to environment variables.

Is this possible in 4.4.2 or later ?

**Reply by flaviuvadan** [reply]
@EricBoix yes! I just tested that. I can also put up an example with this if that's helpful. Here's the example:

```python
from hera.shared import global_config
from hera.workflows import script, Workflow, Parameter, Steps

global_config.host = "..."
global_config.token = ..."
global_config.namespace = "..."

@script(daemon=True)
def daemon() -> None:
    import time 
    while True:  # this needs to stay "alive" for otherwise the IP is not exposed
        time.sleep(10)


@script()
def consume(ip: str) -> None:
    print(ip)


with Workflow(generate_name="daemon-ip-", entrypoint="s") as w:
    with Steps(name="s"):
        d = daemon()
        consume(arguments=Parameter(name="ip", value=d.ip))

w.create()
```

The above ran a workflow with 2 steps:
- first one stays alive because of the while True loop, which basically emulates a server that stays alive and scans for incoming requests
- second one prints the IP of the server that runs in the daemon step

**Comment by EricBoix** [comment]
Thanks  @flaviuvadan . That nails it. I just added an input parameter to the Task/container
requiring the ip number of the server it consumes and then used it inside the `args` field.
Nice.
Closing the discussion.

**Reply by flaviuvadan** [reply]
Very glad to hear it's working for you in that setup @EricBoix!
